lib/components/metrics.js

```diff
@@ -0,0 +1,293 @@
+"use strict";
+
+var fs = require("fs");
+
+var TICKS_PER_SEC = 100; // TODO(paul): look this up via sysconf(_SC_CLK_TCK)
+
+/**
+ * Prometheus-style /metrics gathering and exporting.
+ * This class provides a central location to register gauge and counter metrics
+ * used to generate the <code>/metrics</code> page. It also contains collector
+ * code to generate the standard <code>process_</code> and <code>nodejs_</code>
+ * metrics:
+ *   process_resident_memory_bytes
+ *   process_virtual_memory_bytes
+ *   process_heap_bytes
+ *   process_cpu_seconds_total
+ *   process_open_fds
+ *   process_max_fds
+ *   process_start_time_seconds
+ *   nodejs_heap_used_bytes
+ *
+ * @example <caption>A simple metric that counts the keys in an object:</caption>
+ *   var metrics = new Metrics();
+ *
+ *   var orange = {};
+ *   metrics.addGauge({
+ *       name: "oranges",
+ *       help: "current number of oranges",
+ *       refresh: (gauge) => {
+ *           gauge.set({}, Object.keys(oranges).length);
+ *       },
+ *   });
+ *
+ * @example <caption>Generating values for multiple gauges in a single collector function.</caption>
+ *   var metrics = new Metrics();
+ *
+ *   var oranges_gauge = metrics.addGauge({
+ *       name: "oranges",
+ *       help: "current number of oranges",
+ *   });
+ *   var apples_gauge = metrics.addGauge({
+ *       name: "apples",
+ *       help: "current number of apples",
+ *   });
+ *
+ *   metrics.addCollector(() => {
+ *       var counts = this._countFruit();
+ *       oranges_gauge.set({}, counts.oranges);
+ *       apples_gauge.set({}, counts.apples);
+ *   });
+ *
+ * @example <caption>Using counters</caption>
+ *   var metrics = new Metrics();
+ *
+ *   metrics.addCollector({
+ *       name: "things_made",
+ *       help: "count of things that we have made",
+ *   });
+ *
+ *   function makeThing() {
+ *       metrics.incCounter("things_made");
+ *       return new Thing();
+ *   }
+ *
+ * @constructor
+ */
+function Metrics() {
```

Comment:
Given the strong dependency on Prometheus, I would prefer if this was called `PrometheusMetrics`. The de facto metrics tool in Node is Statsd, so people may naively think that this is actually a statsd metrics class.